From 5f8e74959f19f19018ab0e6400e0f5daba3567e5 Mon Sep 17 00:00:00 2001
From: Jan Chaloupka <jchaloup@redhat.com>
Date: Sat, 26 Nov 2016 13:46:29 +0100
Subject: [PATCH] run etcd on ppc64le by default

---
 etcdmain/etcd.go | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/etcdmain/etcd.go b/etcdmain/etcd.go
index 950513e..7225303 100644
--- a/etcdmain/etcd.go
+++ b/etcdmain/etcd.go
@@ -636,6 +636,14 @@ func checkSupportArch() {
 	if runtime.GOARCH == "amd64" {
 		return
 	}
+
+	if runtime.GOARCH == "ppc64le" {
+		plog.Warningf("Running etcd on ppc64le architecture is not supported.")
+		plog.Warningf("Be aware you are running the etcd at your own risk.")
+		plog.Warningf("For more information visit https://bugzilla.redhat.com/show_bug.cgi?id=1396463")
+		return
+	}
+
 	if env, ok := os.LookupEnv("ETCD_UNSUPPORTED_ARCH"); ok && env == runtime.GOARCH {
 		plog.Warningf("running etcd on unsupported architecture %q since ETCD_UNSUPPORTED_ARCH is set", env)
 		return
-- 
1.9.3

